redfish-core/lib/log_services.hpp

```diff
@@ -6606,4 +6613,581 @@ inline void requestRoutesSystemHardwareIsolationLogService(App& app)
+     * Don't expect log lines to be longer than 4096 so
+     * erroring out to protect against malformed data.
+     */
+    if (!logLine.ends_with('\n'))
+    {
+            /* Line was too long for the buffer.
+             * Try one more read for rest of the line.
+             */
+        if (fgets(buffer.data(), buffer.size(), logStream) != nullptr)
+        {
+            logLine.append(buffer.data());
+        }
+    }
+#endif
+
+    /* Return success only if we got a complete line */
+    return logLine.ends_with('\n');
+}
+
+/**
+ * @brief Reads the audit log entries and writes them to Members array
+ *
+ * @param[in] asyncResp - The redfish response to return.
+ * @param[in] unixfd - File descriptor for Audit Log file
+ * @param[in] skip - Query paramater skip value
+ * @param[in] top - Query paramater top value
+ */
+void readAuditLogEntries(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                         const sdbusplus::message::unix_fd& unixfd, size_t skip,
+                         size_t top)
+{
+    auto fd = dup(unixfd);
+    if (fd == -1)
+    {
+        BMCWEB_LOG_ERROR << "Failed to duplicate fd " << unixfd;
+        messages::internalError(asyncResp->res);
+        return;
+    }
+
+    auto logStream = fdopen(fd, "r");
+    if (logStream == nullptr)
+    {
+        BMCWEB_LOG_ERROR << "Failed to open fd " << fd;
+        messages::internalError(asyncResp->res);
+        close(fd);
+        return;
+    }
+
+    nlohmann::json& logEntryArray = asyncResp->res.jsonValue["Members"];
+    if (logEntryArray.empty())
+    {
+        logEntryArray = nlohmann::json::array();
+    }
+
+    uint64_t entryCount = 0;
+    std::string logLine;
+    while (readLine(logStream, logLine))
+    {
+        /* Note: entryCount counts all entries even ones with parse errors.
+         *       This allows the top/skip semantics to work correctly and a
+         *       consistent count to be returned.
+         */
+        entryCount++;
+        BMCWEB_LOG_DEBUG << entryCount << ":logLine: " << logLine;
+
+        /* Handle paging using skip (number of entries to skip from the
+         * start) and top (number of entries to display).
+         * Don't waste cycles parsing if we are going to skip sending this entry
+         */
+        if (entryCount <= skip || entryCount > skip + top)
+        {
+            BMCWEB_LOG_DEBUG << "Query param skips, line=" << entryCount;
+            continue;
+        }
+
+        nlohmann::json::object_t bmcLogEntry;
+
+        auto auditEntry = nlohmann::json::parse(logLine, nullptr, false);
+
+        LogParseError status = fillAuditLogEntryJson(auditEntry, bmcLogEntry);
+        if (status != LogParseError::success)
+        {
+            BMCWEB_LOG_ERROR << "Failed to parse line=" << entryCount;
+            messages::internalError(asyncResp->res);
+            continue;
+        }
+
+        logEntryArray.push_back(std::move(bmcLogEntry));
+    }
+
```

Comment:
If `readLine()` returns false because the line is too long (i.e. not containing `\n`), should it be handled as an error instead of the end-of-reading?

Perhaps readLine() may need to return the cases - e.g. success, end-of-file, error-in-line?

Reply:
Yes that is a good point. Looking closer at `fgets()` I see that I'm not handling a too-long line correctly. `fgets()` will return up to the size of the buffer in this case, but I've not confirmed that it is an entire line that was read. (My comment in the `#if 0 `section is incorrect.)

I'll rework the `readLine()` function to better handle the error cases.

Reply:
Yes that is a good point. My assumption here is that if one of the lines in the file is longer than 4096 that likely the file itself is corrupt.

However it might just be the case that there is a single very long entry.

I will rework the handling when an entire line is not read to report an error for the specific line but move the SEEK pointer until the end-of-line is actually reached or EOF is reached.

I will continue to have `readLine()` return only true/false with an actual line read on the true case. I will hide the error handling in the function itself.